File: framework-cs/apps/UnityExample/Assets/Samples/Devian Samples/0.1.0/GameContents/Runtime/Ability/AbilityEquip.cs
using Devian.Domain.Game;

namespace Devian
{
    public sealed class AbilityEquip : AbilityBase
    {
        EQUIP mTable = null;
        string mItemUid = string.Empty;
        string mOwnerUnitId = string.Empty;
        int mOwnerSlotNumber = 0;

        public string ItemUid => mItemUid;
        public string EquipId => mTable?.EquipId ?? string.Empty;
        public string OwnerUnitId => mOwnerUnitId;
        public int OwnerSlotNumber => mOwnerSlotNumber;
        public bool IsEquipped => mOwnerSlotNumber > 0;

        public void Init(EQUIP table, string itemUid)
        {
            mTable = table;
            mItemUid = itemUid;
        }

        public override AbilityBase Clone()
        {
            var c = new AbilityEquip();
            c.mTable = mTable;
            c.mItemUid = mItemUid;
            c.mOwnerUnitId = mOwnerUnitId;
            c.mOwnerSlotNumber = mOwnerSlotNumber;
            c.CopyStatsFrom(this);
            return c;
        }

        public void SetOwner(string unitId, int slotNumber)
        {
            mOwnerUnitId = unitId;
            mOwnerSlotNumber = slotNumber;
        }

        public void ClearOwner()
        {
            mOwnerUnitId = string.Empty;
            mOwnerSlotNumber = 0;
        }
    }
}

File: framework-ts/module/devian-domain-game/features/ability/AbilityEquip.ts
import { ITEM_EQUIP } from '../../Generated/Game.g';
import { AbilityBase } from './AbilityBase';

export class AbilityEquip extends AbilityBase {
    private mTable: ITEM_EQUIP | null = null;
    private mItemUid: string = '';
    private mOwnerUnitId: string = '';
    private mOwnerSlotNumber: number = 0;

    get itemUid(): string { return this.mItemUid; }
    get equipId(): string { return this.mTable?.EquipId ?? ''; }
    get ownerUnitId(): string { return this.mOwnerUnitId; }
    get ownerSlotNumber(): number { return this.mOwnerSlotNumber; }
    get isEquipped(): boolean { return this.mOwnerSlotNumber > 0; }

    init(table: ITEM_EQUIP, itemUid: string): void {
        this.mTable = table;
        this.mItemUid = itemUid;
    }

    setOwner(unitId: string, slotNumber: number): void {
        this.mOwnerUnitId = unitId;
        this.mOwnerSlotNumber = slotNumber;
    }

    clearOwner(): void {
        this.mOwnerUnitId = '';
        this.mOwnerSlotNumber = 0;
    }

    clone(): AbilityEquip {
        const c = new AbilityEquip();
        c.mTable = this.mTable;
        c.mItemUid = this.mItemUid;
        c.mOwnerUnitId = this.mOwnerUnitId;
        c.mOwnerSlotNumber = this.mOwnerSlotNumber;
        c.copyStatsFrom(this);
        return c;
    }
}
